src/gluonnlp/utils/parameter.py

```diff
@@ -0,0 +1,113 @@
+# specific language governing permissions and limitations
+# under the License.
+"""Utility functions for parallel processing."""
+
+__all__ = ['clip_grad_global_norm']
+
+import warnings
+
+import numpy as np
+from mxnet import nd
+
+def clip_grad_global_norm(parameters, max_norm, check_isfinite=True):
+    """Rescales gradients of parameters so that the sum of their 2-norm is smaller than `max_norm`.
+    If gradients exist for more than one context for a parameter, these gradients are summed
+    first before calculating the 2-norm.
+
+    .. note::
+
+        This function is only for use when `update_on_kvstore` is set to False in trainer.
+
+    Example::
+
+        trainer = Trainer(net.collect_params(), update_on_kvstore=False, ...)
+        for x, y in mx.gluon.utils.split_and_load(X, [mx.gpu(0), mx.gpu(1)]):
+            with mx.autograd.record():
+                y = net(x)
+                loss = loss_fn(y, label)
+            loss.backward()
+        nlp.utils.clip_grad_global_norm(net.collect_params().values(), max_norm)
+        trainer.update(batch_size)
+        ...
+
+    Parameters
+    ----------
+    parameters : list of Parameters
+    max_norm : float
+    check_isfinite : bool, default True
+         If True, check that the total_norm is finite (not nan or inf). This
+         requires a blocking .asscalar() call.
+
+    Returns
+    -------
+    NDArray or float
+      Total norm. Return type is NDArray of shape (1,) if check_isfinite is
+      False. Otherwise a float is returned.
+
+    """
+    def _reduce_grad(parameter):
+        """Sum gradients of all devices onto the first context and return gradient sum.
+        This function has side effect and does not preserve the original per-device gradients."""
+        grads = parameter.list_grad()
+        if len(grads) > 1:
+            for s in reversed(range(len(grads))):
+                t = int(s/2)
+                if s != t:
+                    target = grads[t]
+                    target[:] = target + grads[s].as_in_context(target.context)
+                else:
+                    continue
+        return grads[0]
+
+    def _broadcast_grad(parameter):
+        """Broadcast the gradient on the first context to the copies on all other contexts."""
+        grads = parameter.list_grad()
+        if len(grads) > 1:
+            for t in range(1, len(grads)):
+                target = grads[t]
+                target[:] = grads[0].as_in_context(target.context)
+
+    def _norm(array):
+        if array.stype == 'default':
+            x = array.reshape((-1,))
+            return nd.dot(x, x)
+        return array.norm().square()
+
+    arrays = [_reduce_grad(p) for p in parameters if p.grad_req != 'null']
+    assert len(arrays) > 0
+    ctx = arrays[0].context
+    total_norm = nd.add_n(*[_norm(arr).as_in_context(ctx) for arr in arrays])
+    total_norm = nd.sqrt(total_norm)
+    if check_isfinite:
+        if not np.isfinite(total_norm.asscalar()):
```

Comment:
We can assign `total_norm = total_norm.asscalar()` if check_isfinite so that it is only done once, then you don't need the if stmt at the end